language/move-prover/src/prelude.bpl

```diff
@@ -963,17 +993,56 @@ procedure {:inline 1} $U64Util_u64_to_bytes(val: Value) returns (res: Value)  {
 // ==================================================================================
 // Native hash
 
-// TODO: implement the below methods
-
-procedure {:inline 1} $Hash_sha2_256(val: Value) returns (res: Value)  {
-    res := DefaultValue;
-    assert false; // $Hash_sha2_256 not implemented
-}
-
-procedure {:inline 1} $Hash_sha3_256(val: Value) returns (res: Value)  {
-    res := DefaultValue;
-    assert false; // $Hash_sha3_256 not implemented
-}
+// Hash is modeled as an otherwise uninterpreted injection.
+// In truth, it is not an injection since the domain has greater cardinality
+// (arbitrary length vectors) than the co-domain (vectors of length 32).  But it is
+// common to assume in code there are no hash collisions in practice.  Fortunately,
+// Boogie is not smart enough to recognized that there is an inconsistency.
+// FIXME: If we had a reliable extensional theory of arrays, and if we could use ==
+// instead of IsEqual, we might be able to avoid so many quantified formulas by
+// using a sha2_inverse function in the ensures conditions of Hash_sha2_256 to
+// assert that sha2/3 are injections without using global quantified axioms.
+
+function $sha2(val: Value) : Value;
+
+// This says that sha2 respects isEquals (this would be automatic if we had an
+// extensional theory of arrays and used ==, which has the substitution property
+// for functions).
+axiom (forall v1,v2: Value :: $Vector_is_well_formed(v1) && $Vector_is_well_formed(v2)
+       && IsEqual(v1, v2) ==> IsEqual($sha2(v1), $sha2(v2)));
+
+// This says that sha2 is an injection
+axiom (forall v1,v2: Value :: $Vector_is_well_formed(v1) && $Vector_is_well_formed(v2)
+	&& IsEqual($sha2(v1), $sha2(v2)) ==> IsEqual(v1, v2));
+
+// This procedure has no body. We want Boogie to just to use its requires
+// and ensures properties when verifying code that calls it.
+procedure $Hash_sha2_256(val: Value) returns (res: Value);
+// It will still work without this, but this helps verifier find more reasonable counterexamples.
```

Comment:
I suppose "this" refers to the commented out precondition.  If the precondition has a benefit, why not add it?

Reply:
The FIXME suggests that there is some provability issue.  Just trying to understand here.

Reply:
The requires was not provable, because the calling context does not guarantee that the "val" argument is vector<u8>. 

I want to understand this better, but I think it's an issue, that, if we fix it, would improve the quality of counterexamples generated by the prover.